// grunt-contrib-download-s3-bucket/tasks/download-s3-bucket.js
/*
 * Copyright (c) 2014 CJ Hanson, contributors
 * Licensed under the MIT license.
 */

'use strict';

var fs = require('fs');
var path = require('path');
var download = require('download');
var AWS = require('aws-sdk');

module.exports = function(grunt) {
    grunt.registerMultiTask('download-s3-bucket', 'Download bucket from Amazon S3', function() {
        // Merge task-specific and/or target-specific options with these defaults.
        var options = this.options({
            bucket: ''
        });

        console.log(options);

        var async = this.async;

        var done = async();
        var taskCount = 0;
        var allTasksOK = true;
        var addTask = function(){
            ++taskCount;
        }
        var allDone = function(allOK){
            if(!allOK)
                allTasksOK = false;

            --taskCount;

            if(taskCount == 0){
                console.log('All done '+allTasksOK);
                done(allTasksOK);
            }
        }

        AWS.config.loadFromPath('./awsconfig.json');

        console.log(AWS.config.credentials);

        var s3 = new AWS.S3();

        var params = {
            Bucket: options.bucket, // required
            Delimiter: '',
            EncodingType: 'url',
            Marker: '',
            MaxKeys: 0,
            Prefix: ''
        };

        addTask();

        s3.listObjects({Bucket: options.bucket}, function(err, data) {
            if (err) {
                console.log(err, err.stack);
                allDone(false);
            }else {
                if(data.IsTruncated){
                    //TODO request more
                }

                for(var i = 0; i < data.Contents.length; ++i){
                    addTask();
                }

                data.Contents.forEach(function(element, index, array){
                    console.log(element.Key);
                    if(path.extname(element.Key) == ''){
                        if(!fs.existsSync(element.Key)) {
                            console.log('mkdir ' + element.Key);
                            fs.mkdirSync(element.Key);
                        }
                        allDone(true);
                    }else{
                        var url = 'https://s3.amazonaws.com/' + options.bucket + '/' + element.Key;
                        var dl = download(url, element.Key);

                        dl.on('close', function () {
                            console.log('DL complete '+element.Key);
                            allDone(true);
                        });

                        dl.on('error', function (status) {
                            console.log('DL error '+status+' '+element.Key);
                            allDone(false);
                        });
                    }
                });

                //console.log(data);
                allDone(true);
            }
        });
    });
};
